### PR TITLE
Fix switch-in abilities activating on an empty field

### LINKED ISSUE #2371: Switch-in abilities activate against fainted mons
Example: 
Current behaviour:
1. Player's mon KO's an AI mon with Volt Switch and switches out
2. Player sends out a mon with Intimidate
3. Intimidate activates against the fainted mon
4. AI sends out its next mon

Expected behaviour:
1. Player's mon KO's an AI mon with Volt Switch and switches out
2. Player sends out a mon with Intimidate
3. AI sends out its next mon
4. Intimidate activates against the new opponent

Presumably this also affects Trace, Frisk, Download, Anticipation, Forewarn and whatever other switch in abilities interact with the opponents.

### LINKED ISSUE #2619: Some abilities are broken in certain battle types like battle factory
Have an enemy mon
Ko it
Enemy trainer sends a mon with Intimidate
Popup gets enabled from player side and not from enemies, including wrong buffers on the text.

SBird made a test for this and it's confirmed to be broken.

### PR DESCRIPTION
**Pull request by DizzyEggg**
Fixes #2619
Fixes #2371

Takes care of PRs #2622 and #2148.

Added tests for Download and Intimidate. 
SwitchInDownload
intimidate

### CODE REVIEW

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -6900,6 +6900,35 @@ static void SetDmgHazardsBattlescript(u8 battlerId, u8 multistringId)
...
+             || (gFieldStatuses & STATUS_FIELD_TERRAIN_ANY && AbilityBattleEffects(ABILITYEFFECT_ON_TERRAIN, battlerId, 0, 0, 0))
+             || ItemBattleEffects(ITEMEFFECT_ON_SWITCH_IN, battlerId, FALSE)
+             || AbilityBattleEffects(ABILITYEFFECT_TRACE2, 0, 0, 0, 0));
+}
+
+bool32 ShouldPostponeSwitchInAbilities(u32 battlerId)
+{
+    bool32 aliveOpposing1 = IsBattlerAlive(BATTLE_OPPOSITE(battlerId));
+    bool32 aliveOpposing2 = IsBattlerAlive(BATTLE_PARTNER(BATTLE_OPPOSITE(battlerId)));
+    // No pokemon on opposing side - postopone.
```
Comment (LOuroboros) [inline src/battle_script_commands.c:6916]:
```suggestion
    // No pokemon on opposing side - postpone.
```

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -7016,12 +7045,17 @@ static void Cmd_switchineffects(void)
...
         gDisableStructs[gActiveBattler].truantSwitchInHack = 0;
 
-        if (AbilityBattleEffects(ABILITYEFFECT_ON_SWITCHIN, gActiveBattler, 0, 0, 0)
-         || (gBattleWeather & B_WEATHER_ANY && WEATHER_HAS_EFFECT && AbilityBattleEffects(ABILITYEFFECT_ON_WEATHER, gActiveBattler, 0, 0, 0))
-         || (gFieldStatuses & STATUS_FIELD_TERRAIN_ANY && AbilityBattleEffects(ABILITYEFFECT_ON_TERRAIN, gActiveBattler, 0, 0, 0))
-         || ItemBattleEffects(ITEMEFFECT_ON_SWITCH_IN, gActiveBattler, FALSE)
-         || AbilityBattleEffects(ABILITYEFFECT_TRACE2, 0, 0, 0, 0))
-            return;
+        // Don't activate switch-in abilities if the opposing field is empty.
+        // This could happen when a mon uses explosion and causes everyone to faint.
```
Comment (LOuroboros) [inline src/battle_script_commands.c:7049]: Then maybe, now that we have this, we wouldn't need to keep the `VARIOUS_JUMP_IF_NO_VALID_TARGETS` which I explicitly added to solve the interaction with Explosion? :eyes:

File: test/ability_download.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
+#include "global.h"
+#include "test_battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gBattleMoves[MOVE_TACKLE].split == SPLIT_PHYSICAL);
+    ASSUME(gBattleMoves[MOVE_TRI_ATTACK].split == SPLIT_PHYSICAL);
```
Comment (mrgriffin) [inline test/ability_download.c:7]: :skull:

Comment (DizzyEggg) [reply test/ability_download.c]: Fixed

File: src/battle_script_commands.c (hunk 2 above)
Comment (DizzyEggg) [reply src/battle_script_commands.c:7068]: Done, I removed VARIOUS_JUMP_IF_NO_VALID_TARGETS

**Review by mrgriffin — APPROVED** [review]

**Pull request merged**